### DISCUSSION TITLE
Selenium動態爬蟲

### DISCUSSION THREAD

**Discussion by ATREE01** [1101-GS4719-C/Python-Community-GS4719-C#1666] [General]
#    Discord bot + Selenium動態爬蟲
##    前言
這篇文章會教你實作一個爬蟲讓你可以在Discord內使用指令讓機器人把該篇貼文中的圖片發送到Discord中。

#    本文
## 動態爬蟲
今天我們會用的一個跟靜態爬蟲不一樣的套件叫做`Selenium`。

這個套件跟request不一樣的地方在於它可以模擬使用者在使用網頁的過程，讓程式碼去跟網頁作互動，像是點擊一些按鈕或是上下滑動等等。

因為Selenium需要選擇你要用的瀏覽器去下載對應的driver的關係，這篇文章將會以Chrome示範[下載網址](https://googlechromelabs.github.io/chrome-for-testing/)，在這個網頁裡面選擇跟你的Chrome主要版本相同的Win64下載
![](https://hackmd.io/_uploads/ryC6_gpba.png)

網頁版本在這邊看的到。

![](https://hackmd.io/_uploads/SJN7dl6-p.png)

下載下來的壓縮黨裡面會有這兩個東西，把`chromedriver.exe`移動到跟你的python檔案的相同資料夾當中。
![](https://hackmd.io/_uploads/H1fZFga-p.png)

執行這段程式碼之後會看到一個網頁被打開

```python
from selenium import webdriver
from selenium.webdriver.chrome.options import Options #之後可以自訂一些設定
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By #至後用來尋找元素時會用到

driver = webdriver.Chrome(service=Service('chromedriver.exe'))

url = "https://www.instagram.com/p/CydwI_-oP1V/?utm_source=ig_web_copy_link&igshid=MzRlODBiNWFlZA%3D%3D"

driver.get(url)
```
在讀取完之後會看到以下畫面，代表你順利地透過Selenium來打開網頁了。

![](https://hackmd.io/_uploads/SJFH3eaZp.png)

如果因為電腦讀取速度比較慢的關係網頁還沒有跑出來就被關掉了可以加上下方程式碼當中的兩行，讓你的程式睡一下給網頁一些時間讀取。

```python
from selenium import webdriver
from selenium.webdriver.chrome.options import Options #之後可以自訂一些設定
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By #至後用來尋找元素時會用到

import time #加入這行

driver = webdriver.Chrome(service=Service('chromedriver.exe'))

url = "https://www.instagram.com/p/CydwI_-oP1V/?utm_source=ig_web_copy_link&igshid=MzRlODBiNWFlZA%3D%3D"

driver.get(url)
time.sleep(10) #加入這行
```

接下來因為我們想要把圖片傳送到Discord中，我們需要先把圖片下載下來，所以我們要先來分析一下這篇貼文的HTML。

先手動打開連結，記得這邊要使用**無痕模式**，因為Instagram在登入狀態下跟未登入狀態下的class名稱是不一樣的而我們的Selenium是在未登入的狀態下去操作的。

![](https://hackmd.io/_uploads/rJQIClabp.png)

找到這篇貼文的class`x5yr21d xu96u03 x10l6tqk x13vifvy x87ps6o xh8yej3`這邊要注意一件事情，因為有時會同時有很多相同class的元素，因此有時要確認用來搜尋的class能不能順利找到我們要找的東西，不然的話就要考慮用Xpath來做尋找。而這邊因為我們要尋找的圖片剛好是相同class當中的第一個，因次可以放心使用。

```python=
driver.get(url)
img = driver.find_element(By.CLASS_NAME, "x5yr21d.xu96u03.x10l6tqk.x13vifvy.x87ps6o.xh8yej3")
src = img.get_attribute('src')

# print(src) 可以把src輸出出來看是不是你要找的東西。
```

在確認成功拿到圖片的連結之後我們就可以使用上次教過的request把圖片下載下來。

```python
img = requests.get(src) # 記得要先 import requests
with open("img.png", "wb") as file: # wb 是寫入二進位模式
    file.write(img.content)
```
執行完這段之後就會看到img.png出現在你的資料夾當中了。

到這邊為止我們還沒有真正利用到甚麼動態爬蟲的功能。這是因為這邊的圖片只有一張，如果有多張圖片的話就會需要用到動態爬蟲的功能。

範例網址: https://www.instagram.com/p/CyLR_BHxiKB/?img_index=1

這是一個有多張圖片的貼文。我們先觀察一下HTML

![](https://hackmd.io/_uploads/BJLY4bTWp.png)

HTML解釋
`ul`:代表這是一個 沒有順序關係的List
`li`:代表的是`ul`這個list中的元素
我們可以看到雖然這篇貼文有10張圖片，但是這邊的`li`只有兩個但是這時候如果我們按下向右的按鈕會發現有新的`li`出現。可是多按幾次之後會發現前面的`li`又不見了。這邊就是我們可以用到動態爬蟲的地方了。

要下載多張圖片我們可以做以下操作:

當我們打開一篇貼文的時候，我們嘗試去找有沒有向右的按鈕，如果有的話接下來我們先按照上面的步驟找到圖片的網址，把網址存下來後點擊向右的按鈕。重複這段步驟我們就可以把所有的圖片網址都存下來，接下來只要下載就好。

這邊我為了兩單張圖片跟多張圖片的操作可以比較一致我們在觀察一下會發現不管是單張圖片或是多張圖片，整個圖片的區域都是被包在`clss=_aatk _aatn`的`div`裡面因此我們一開始用`post = driver.find_element(By.CLASS_NAME, "_aatk._aatn")`去取得這個區塊。

![](https://hackmd.io/_uploads/Sy6vQzpbT.png)

```python

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By

import time

import requests

chrome_options = Options()
#chrome_options.add_argument('--headless') 可以讓網頁不會出現 只在背景運作
#chrome_options.add_argument('--log-level=3') 不顯示網頁之後會出現一些不重要警告或是錯誤 ，不想讓它顯示的話加上這行。
driver = webdriver.Chrome(service=Service('chromedriver.exe'),options = chrome_options)

url = "https://www.instagram.com/p/CyLR_BHxiKB/?img_index=1"

driver.get(url)
time.sleep(5)#這邊這個睡眠時間取決於你的電腦，可以試試看短一點，如果發現網頁讀取不出來的話可以弄長一點。
    
post = driver.find_element(By.CLASS_NAME, "_aatk._aatn")#取得區塊
img_srcs = []#用來存連結的list
try:
    while True:
        button = driver.find_element(By.CLASS_NAME, "_afxw._al46._al47") #試著尋找有沒有button如果沒找地的話會跳exception 接下來就只會執行except的部分
        datas = post.find_elements(By.CLASS_NAME, "x5yr21d.xu96u03.x10l6tqk.x13vifvy.x87ps6o.xh8yej3") #找到所有的 注意這邊是find_elements 後面有s
        for data in datas:#遍歷所有找到的div
            img_src = data.get_attribute('src')#取得他的src屬性
            if img_src not in img_srcs: #如果這個連結之前沒有儲存過
                img_srcs.append(img_src)
        button.click()#最重要的部分，去點擊那個向右的按鈕

except:#單張圖片的做法
    data = post.find_element(By.CLASS_NAME, "x5yr21d.xu96u03.x10l6tqk.x13vifvy.x87ps6o.xh8yej3")
    img_srcs.append(data.get_attribute('src'))
    
for index, url in enumerate(img_srcs):
    response = requests.get(url)
    with open(f'image{index}.png', 'wb') as file:
        file.write(response.content)
```

這個就是完整的程式碼了，如果有甚麼不懂的地方可以提問。

礙於篇幅的關係，機器人的部分下次再說Bye Bye~。

**Comment by 1chooo** [comment]
這個動態爬蟲的說明文章真的非常詳細，之前也用過 `Selenium` 做氣象資料的動態調整，都在 Debug 模擬操作電腦，期待你下一篇機器人的部分！
